storage_scrubber/src/find_large_objects.rs

```diff
@@ -0,0 +1,97 @@
+use futures::StreamExt;
+use pageserver::tenant::storage_layer::{DeltaLayerName, ImageLayerName};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    init_remote, list_objects_with_retries, metadata_stream::stream_tenants, BucketConfig, NodeKind,
+};
+
+#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
+enum LargeObjectKind {
+    DeltaLayer,
+    ImageLayer,
+    Other,
+}
+
+impl LargeObjectKind {
+    fn from_key(key: &str) -> Self {
+        let fname = key.split('/').last().unwrap();
+        let is_delta = DeltaLayerName::parse_str(fname).is_some();
+        let is_image = ImageLayerName::parse_str(fname).is_some();
+        match (is_delta, is_image) {
+            (true, false) => LargeObjectKind::DeltaLayer,
+            (false, true) => LargeObjectKind::ImageLayer,
+            (false, false) | (true, true) => LargeObjectKind::Other,
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct LargeObject {
+    pub key: String,
+    pub size: u64,
+    kind: LargeObjectKind,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct LargeObjectListing {
+    pub objects: Vec<LargeObject>,
+}
+
+pub async fn find_large_objects(
+    bucket_config: BucketConfig,
+    min_size: u64,
+    ignore_deltas: bool,
+) -> anyhow::Result<LargeObjectListing> {
+    let (s3_client, target) = init_remote(bucket_config.clone(), NodeKind::Pageserver)?;
+    let mut tenants = std::pin::pin!(stream_tenants(&s3_client, &target));
+    let mut objects = Vec::new();
+    let mut tenant_ctr = 0u64;
+    let mut object_ctr = 0u64;
+    while let Some(tenant_shard_id) = tenants.next().await {
+        let tenant_shard_id = tenant_shard_id?;
+        let mut tenant_root = target.tenant_root(&tenant_shard_id);
+        // We want the objects and not just common prefixes
+        tenant_root.delimiter.clear();
+        let mut continuation_token = None;
```

Comment:
I don't have much experience with the scrubber code base but needing to reach into the struct like this seems like an anti-pattern to me. What happens if you don't do this?

Reply:
Same is done in `get_tenant_objects` and `get_timeline_objects`: https://github.com/neondatabase/neon/blob/27518676d7aebb26ad81bf0a926749c5ed9e75d5/storage_scrubber/src/garbage.rs#L339 . If I don't do this, then only the prefixes are returned, but the intent is to do a recursive listing of the tenant prefix.

At the start I've done it "properly" to list each timeline and then inside each timeline, list each object. Then this isn't needed, but it is a) more code to add a timeline loop around the object loop, and b) much slower, as then each timeline would issue another list API call, bringing down the average of # of objects returned by an API call. You can see the performance impact in [this](https://github.com/neondatabase/neon/pull/8257/commits/a36c679eb816424e152acaf976ca26d7eb04cc0a) commit message.

In theory I could probably just do one big list call, it would be even faster I suppose, but then it's harder to do per-tenant statistics.

Reply:
Hm, so, why not just use `get_tenant_objects` ? Maybe factor out a `get_tenant_objects_stream` so we can avoid the `.collect()` ?

Anyway, I don't feel too strongly about this, so, feel free to resolve. Seems like it's common practice in the code base already.

Reply:
Ah ok, ObjectIdentifier, I see.